### PR TITLE
#164046210: users can log in

### PR DESCRIPTION
**Pull request by MandelaK**
### What does this PR do?
Add login functionality to the application. When a user fills in the login form, they should get an appropriate response. Error messages are displayed in red and form validation is done when the form is submitted.

### Description of task
Users should be able to log in so they can be identified and have a custom experience. When a user logs in, we store the token and dispatch an action to the redux store. A helper function is implemented to ensure that only valid tokens can be decoded.

The following are saved to localStorage when a user logs in:
    - `profileImage` - Link to the user's profile image
    - `token` - The user's token
    - `email` - The user's email
    - `username` - The username of the user

### How has this been tested?
- [x] End-to-end testing
- [x] Unit tests

### How can this be manually tested?
- Clone this repo and checkout to the branch `ft-users-login-#164046210`
- run npm install and create a .env file in your environment variables. Add the variable `REACT_APP_BASE_URL` to the env file and have it's value point to `https://ah-legion-staging.herokuapp.com/api/` or your custom backend. Also add a `REACT_APP_SECRET_KEY` which should be the same as that on heroku.
- Go to the login page and try logging in a user 

### Other information
- I have implemented a function in the `utils` that checks whether a user is logged in by attempting to authenticate the token stored in the `localStorage`
- Make sure you have run `npm install` and properly configure environment variables. Also restart your server and run `source .env`.
- You can use `axiosConfig` to make API calls

### PT
[164046210](https://www.pivotaltracker.com/story/show/164046210)

### CODE REVIEW

File: src/store/reducers/index.js [hunk 1]
```diff
@@ -1,12 +1,6 @@
...
+import { combineReducers } from 'redux';
+import loginReducer from './LoginReducer';
 
-function rootReducer(state = initialState, action){
-
-    return state;
-};
-
-export default rootReducer;
+export default combineReducers({
```
Comment (ElMonstro) [inline src/store/reducers/index.js]: Can I suggest this, It is more readable:
```suggestion
const rootReducer = combineReducers({
login_user: loginReducer})
export default rootReducer
```

**Comment by abulojoshua1** [comment]
@MandelaK, kindly rebase so as to add coveralls to your PR.

File: src/containers/LoginForm.jsx [hunk 2]
```diff
@@ -0,0 +1,155 @@
+import React from "react";
```
[the remark this answers is not in the document]
Comment (ElMonstro) [reply src/containers/LoginForm.jsx]: Please follow this recommendation

File: src/tests/store/Actions/authActions/loginAction.test.js [hunk 3]
```diff
@@ -0,0 +1,71 @@
+import axiosconfig from "../../../../axiosConfig";
```
Comment (ElMonstro) [inline src/tests/store/Actions/authActions/loginAction.test.js]: This does not conform to camelCase naming conventions, consider axiosConfig

**Comment by ElMonstro** [comment]
@MandelaK Please fix the linting errors

File: src/index.jsx [hunk 4]
```diff
@@ -8,6 +8,8 @@ import Routes from './routes';
 import * as serviceWorker from './serviceWorker';
 import store from './store/store';
 
+require('dotenv').config();
```
Comment (cob04) [inline src/index.jsx:12]: Would you check if this is really necessary? I'm able to to use `dotenv` without this.

File: src/components/Header.jsx [hunk 5]
```diff
@@ -5,31 +5,52 @@ import { NavLink } from 'react-router-dom';
 import AuthenticationModal from './AuthenticationModal';
 import store from '../store/store';
 import { LOGIN, REGISTER } from '../store/actions/actionTypes';
-
+import { isLoggedIn } from '../utils/tokenValidator';
 
 class Header extends React.Component {
+
+  state = {
```
[the remark this answers is not in the document]
Comment (ElMonstro) [reply src/components/Header.jsx:11]: Initialize state in a constructor to remove this error @MandelaK

File: src/containers/LoginForm.jsx [hunk 6]
```diff
@@ -0,0 +1,152 @@
...
+import PropTypes from 'prop-types';
+import { connect } from 'react-redux';
+import loginAction from '../store/actions/authActions/LoginAction';
+import validateLoginForm from '../utils/validation';
+import store from '../store/store';
+import { REGISTER, LOGIN_LOADING } from '../store/actions/actionTypes';
+
+export class LoginForm extends React.Component {
+  static propTypes = {
+    loginUser: PropTypes.object.isRequired,
```
Comment (ElMonstro) [inline src/containers/LoginForm.jsx]: Do this to eliminate this error @MandelaK 
```suggestion
    loginUser: PropTypes.shape({}).isRequired,,
```

**Review by Etomovich — APPROVED** [review]

File: src/index.jsx (hunk 4 above)
Comment (MandelaK) [reply src/index.jsx:12]: I don't if it's necessary, but the documentation insist you do it ...

**Review by RachelleMaina — APPROVED** [review]
This looks good.

**Review by BrianSerem — APPROVED** [review]

**Review by cob04 — APPROVED** [review]
Looks OK

**Review by ElMonstro — APPROVED** [review]
Approved

**Comment by ElMonstro** [comment]
Please fix the hound format @MandelaK

File: src/tests/store/Actions/authActions/loginAction.test.js (hunk 3 above)
Comment (MandelaK) [reply src/tests/store/Actions/authActions/loginAction.test.js]: Thanks, it was actually a typo on my part

File: src/components/Header.jsx (hunk 5 above)
[the remark this answers is not in the document]
Comment (MandelaK) [reply src/components/Header.jsx:11]: I used the babel-parser .... this is no longer an issue. Thanks for the suggestion though

File: src/css/App.css [hunk 7]
```diff
@@ -150,3 +150,76 @@ body {
...
+	background-color: transparent;
+	border: 0;
+	-webkit-appearance: none;
+	-moz-appearance: none;
+	appearance: none;
+}
+
+.alert {
+  text-align: center;
+  /* border: 0; */
```
Comment (abulojoshua1) [inline src/css/App.css]: @MandelaK, kindly take out this commented out code.

File: src/containers/LoginForm.jsx [hunk 8]
```diff
@@ -0,0 +1,166 @@
...
+              <Spinner
+                as="span"
+                animation="grow"
+                size="sm"
+                role="status"
+                aria-hidden="true"
+                className="btn-disabled"
+              />
+              Loading...
+            </Button>
```
Comment (abulojoshua1) [inline src/containers/LoginForm.jsx]: @MandelaK, you can do this in a cleaner way, sync with @ElMonstro to help you out on this.

File: src/containers/LoginForm.jsx [hunk 9]
```diff
@@ -0,0 +1,166 @@
...
+} from 'react-bootstrap';
+import PropTypes from 'prop-types';
+import { connect } from 'react-redux';
+import loginAction from '../store/actions/authActions/LoginAction';
+import validateLoginForm from '../utils/validation';
+import store from '../store/store';
+import { REGISTER, IS_LOADING } from '../store/actions/actionTypes';
+
+export class LoginForm extends React.Component {
+  static propTypes = {
```
Comment (abulojoshua1) [inline src/containers/LoginForm.jsx]: @MandelaK, place this proptypes at the bottom of the render function.

File: src/containers/LoginForm.jsx [hunk 10]
```diff
@@ -0,0 +1,166 @@
...
+  constructor(props) {
+    super(props);
+    this.state = {
+      email: '',
+      password: '',
+      formErrors: {},
+      touched: false,
+      show: true,
+    };
+    this.onAlertClose = this.onAlertClose.bind(this);
```
Comment (abulojoshua1) [inline src/containers/LoginForm.jsx]: @MandelaK, use arrow functions in instead of making a bind statement.

File: src/containers/LoginForm.jsx [hunk 11]
```diff
@@ -0,0 +1,166 @@
...
+    const { email, password } = this.state;
+    event.preventDefault();
+    const formErrors = validateLoginForm(email, password);
+    if (formErrors.length === 0) {
+      this.setState({
+        formErrors: [],
+      });
+      const data = { email, password };
+      store.dispatch({ type: IS_LOADING, payload: true });
+      this.props.loginAction(data);
```
Comment (abulojoshua1) [inline src/containers/LoginForm.jsx]: @MandelaK, kindly destructure props at the top of the class method.

**Review by abulojoshua1 — CHANGES_REQUESTED** [review]

**Comment by abulojoshua1** [comment]
@MandelaK], kindly fix the merge conflicts.

**Review by abulojoshua1 — APPROVED** [review]
LGTM

**Pull request merged**